### PR TITLE
FIX PVeRA forward implementation for bitsandbytes

### PR DESCRIPTION
**Pull request by BenjaminBossan**
In the forward path of PVeRA, a sampling step is required. For the bnb layers, this was missing. The PR now adds the sampling step.

I confirmed locally that this fixes the [failing PVeRA tests in the nightly CI](https://github.com/huggingface/peft/actions/runs/24814323261/job/72625376879#step:6:1482). Failing tests on CPU CI are unrelated.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
@leofillioux Please review.

File: src/peft/tuners/pvera/bnb.py [hunk 1]
```diff
@@ -223,9 +225,9 @@ def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
                     sliced_B = pvera_B[: self.out_features, :].to(x.device)
 
                     x_temp = dropout(x.to(lambda_d.dtype))
-
-                    adapter_output = lambda_b * torch.nn.functional.linear(
-                        lambda_d * torch.nn.functional.linear(x_temp, sliced_A), sliced_B
+                    mu, logvar = (lambda_d * F.linear(dropout(x_temp), sliced_A)).chunk(2, dim=-1)
+                    adapter_output = lambda_b * F.linear(
+                        self._reparametrize(mu, logvar, self.sample_at_inference), sliced_B
```
Comment (BenjaminBossan) [inline src/peft/tuners/pvera/bnb.py:230]: Now mirrors:

https://github.com/huggingface/peft/blob/76c37d4686cf7245a9a6998fd1e2619d62c9322d/src/peft/tuners/pvera/layer.py#L300-L303

File: src/peft/tuners/pvera/bnb.py [hunk 2]
```diff
@@ -223,9 +225,9 @@ def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
                     sliced_B = pvera_B[: self.out_features, :].to(x.device)
 
                     x_temp = dropout(x.to(lambda_d.dtype))
-
-                    adapter_output = lambda_b * torch.nn.functional.linear(
-                        lambda_d * torch.nn.functional.linear(x_temp, sliced_A), sliced_B
+                    mu, logvar = (lambda_d * F.linear(dropout(x_temp), sliced_A)).chunk(2, dim=-1)
```
Comment (leofillioux) [inline src/peft/tuners/pvera/bnb.py]: Dropout if already applied above `x_temp = dropout(x.to(lambda_d.dtype))`. I think we should just use `x_temp` directly in the linear layer here.

File: src/peft/tuners/pvera/bnb.py [hunk 3]
```diff
@@ -392,9 +395,9 @@ def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
...
```
Comment (leofillioux) [inline src/peft/tuners/pvera/bnb.py]: Same comment on the double use of dropout as above.

**Review by leofillioux — COMMENTED** [review]
Hey @BenjaminBossan! Almost everything looked good, except for the double use of dropout.

**Comment by BenjaminBossan** [comment]
@leofillioux Good catch, that was indeed an error on my part. It should be fixed now, please check again. (failing CI is still unrelated)

**Review by leofillioux — COMMENTED** [review]
Everything looks good to me! Thank you!

File: src/peft/tuners/pvera/layer.py [hunk 4]
```diff
@@ -143,6 +143,15 @@ def reset_pvera_parameters(self, adapter_name, d_initial: float = 0.1):
                 nn.init.zeros_(self.pvera_lambda_d[adapter_name]).fill_(d_initial)
                 nn.init.zeros_(self.pvera_lambda_b[adapter_name])
 
+    def _reparametrize(self, mu, logvar, sample_at_inference):
+        if self.training or (not self.training and sample_at_inference):
+            std = torch.exp(0.5 * logvar)
+            eps = torch.randn_like(std)
```
Comment (sayakpaul) [inline src/peft/tuners/pvera/layer.py:149]: Thinking out loud. Should there be a `generator` argument to control stochasticity?

**Review by sayakpaul — APPROVED** [review]
Thanks!

File: src/peft/tuners/pvera/layer.py (hunk 4 above)
Comment (BenjaminBossan) [reply src/peft/tuners/pvera/layer.py:149]: Good point, I agree that having an option to seed would be good here. @leofillioux WDYT? I still merged this PR, as it is a separate issue.

Comment (leofillioux) [reply src/peft/tuners/pvera/layer.py:149]: Yes that's a good idea! I made an implementation, and I'm running the tests. I'll open a PR once they're done.

Comment (leofillioux) [reply src/peft/tuners/pvera/layer.py:149]: @BenjaminBossan I still have problems with my environment and I'm not able to run the tests :/ Should I still make a PR?

Comment (BenjaminBossan) [reply src/peft/tuners/pvera/layer.py:149]: Sure, go ahead.

Comment (leofillioux) [reply src/peft/tuners/pvera/layer.py:149]: I opened a PR (#3200) to address this problem.
